File: demo/go/generic/join.go
package generic

type Cond[L, R any] struct {
	// LF keyof L
	// RF keyof R
}

// cond出现的字段必须分别在L和R均出现
// 类型和接口可以，但是类型的字段不行，粒度不够细
func leftJoin[L any, R any, LR any](
	left []L,
	right []R,
	cond Cond[L, R],
	f func(l L, r R) LR,
) []LR {

	return nil
}

// 缩窄范围，条件均使用指定字段Id，其类型为comparable
type IdConstraint[T comparable] interface {
	~struct {
		Id T

		// 怎么表示还有其它任意的字段呢
		Data any
	}
}

func leftJoinById[T comparable, L, R IdConstraint[T], LR any](left []L, right []R, f func(l L, r R) LR) []LR {
	// rm := make(map[T]R)
	// for _, one := range right {
	// 	rm[one.Id] = one // 暂未支持，1.19有望加上
	// }

	res := make([]LR, 0, len(left))
	// for _, one := range left {
	// 	r := rm[one.Id]
	// 	lr := f(one, r)
	// 	res = append(res, lr)
	// }

	return res
}

File: demo/ts/join.ts

// Cond出现的字段必须分别在L和R均出现
export type Cond<L, R> = {
    l: keyof L
    r: keyof R
}

export function leftJoin<L, R, LR>(
    left: L[],
    right: R[],
    cond: Cond<L, R>,
    f: (l: L, r: R) => LR,
): LR[] {
    const rm = new Map()
    right.forEach((value: R, _index: number, _array: R[]) => {
        const rv = getProperty(value, cond.r)
        rm.set(rv, value)
    })

    const res = new Array(left.length)
    left.forEach((value: L, _index: number, _array: L[]) => {
        const lv = getProperty(value, cond.l)
        const rv = rm.get(lv)
        const lr = f(value, rv)
        res.push(lr)
    })

    return res
}

export function getProperty<T, K extends keyof T>(o: T, propertyName: K): T[K] {
    return o[propertyName]; // o[propertyName] is of type T[K]
}

const users = [
    {
        id: 1,
        name: "jd",
    },
    {
        id: 2,
        name: "jk",
    },
]

const articles = [
    {
        id: 1,
        userId: 1,
        name: "join"
    },
    {
        id: 2,
        userId: 2,
        name: "join 2"
    },
]

const res = leftJoin(users, articles, { l: "id", r: "userId" }, (l, r) => {
    return {
        id: r.id,
        name: r.name,
        userId: l.id,
        userName: l.name,
    }
})
console.log("res: ", res)
